File: frontend/apps/epicerie/src/types/etl_import.ts
// Types ETL Import — Queue factures fournisseur (ADR-25)

export type AutoFillSource =
  | 'norm_exact'        // S1 : désignation normalisée identique
  | 'article_four'      // S2 : même code article fournisseur
  | 'correction_history' // S3 : correction manuelle historique
  | 'tfidf_auto'        // S4 : TF-IDF ≥ 0.90 (silencieux)
  | 'tfidf_auto_soft'   // S4 : TF-IDF 0.80-0.90 (badge visible)
  | 'off_import'        // S6 : import catalogue OpenFoodFacts France

export interface AutoAppliedFieldMeta {
  source: AutoFillSource
  score: number
  candidate_id?: number | null
  candidate_designation?: string | null
}

export interface SimilarProduct {
  candidate_id: number
  designation?: string
  ean?: string | null
  categorie_code?: string | null
  marque?: string | null
  conditionnement?: string | null
  volume_unitaire_ml?: number | null
  prix_unitaire_cts?: number | null
  taux_tva_centieme?: number | null
  score: number
}

export type EtlImportStatut =
  | 'PENDING'
  | 'RUNNING'
  | 'PREVIEW'
  | 'VALIDATED'
  | 'REJECTED'
  | 'SUCCES'
  | 'PARTIEL'
  | 'ECHEC'
  | 'REVERTED'

export interface EtlImportRead {
  id: number
  fournisseur_id: number | null
  fichier_source: string | null
  fichier_path: string | null
  statut: EtlImportStatut
  validation_step?: string | null
  nb_lignes_total: number | null
  nb_lignes_ok: number
  nb_lignes_conflit: number
  nb_lignes_erreur: number
  erreur_detail: string | null
  numero_facture: string | null
  date_facture: string | null
  montant_ht_total: number | null
  montant_tva_total: number | null
  montant_ttc_total: number | null
  vendor_code: string | null
  quality_score: number | null
  ecart_reconciliation: number | null
  client_name: string | null
  target_tenant_id: number | null
  created_at: string
  updated_at: string
}

export interface EtlImportListResponse {
  items: EtlImportRead[]
  total: number
}

export interface FactureUploadResponse {
  etl_import_id: number
  statut: string
  nb_lignes: number
  numero_facture: string | null
  vendor_code: string | null
  quality_score: number | null
}

export interface EtlImportUpdateRequest {
  numero_facture?: string
  date_facture?: string
  montant_ht_total?: number
  montant_tva_total?: number
  montant_ttc_total?: number
}

export interface LigneFactureRead {
  idx: number
  designation: string
  designation_raw?: string | null
  ean?: string | null
  article_fournisseur?: string | null
  marque?: string | null
  categorie_code?: string | null
  conditionnement?: string | null
  unite_base: string
  source_fournisseur: string
  quantite?: number | null
  colisage?: number | null
  prix_unitaire_cts?: number | null
  montant_ht_cts?: number | null
  montant_ttc_cts?: number | null
  taux_tva_centieme?: number | null
  degre_alcool?: number | null
  contenant?: string | null
  volume_unitaire_ml?: number | null
  // Coordonnées PDF et confiance (Phase 2)
  page_number?: number | null
  y_position?: number | null
  confidence_score?: number | null
  // Produits similaires et prix catalogue (Phase 2+3)
  similar_products?: SimilarProduct[] | null
  // Auto-fill traçable (S1-S5)
  auto_applied_fields?: Record<string, AutoAppliedFieldMeta> | null
  prix_catalogue_actuel_cts?: number | null
}

export interface EtlImportDetail extends EtlImportRead {
  lignes: LigneFactureRead[]
  montant_ht_calcule: number | null
  montant_ttc_calcule: number | null
}

// ── Ligne mutations ──────────────────────────────────────────────────────────

export interface LigneFactureUpdate {
  idx: number
  designation?: string
  ean?: string
  marque?: string
  conditionnement?: string
  categorie_code?: string
  quantite?: number
  prix_unitaire_cts?: number
  taux_tva_centieme?: number
  auto_applied_fields?: Record<string, AutoAppliedFieldMeta> | null
}

export interface LignesBatchUpdateRequest {
  updates: LigneFactureUpdate[]
}

// ── Édition post-validation (Option B — P1/P2) ─────────────────────────────

export interface ValidatedLigneEdit {
  idx: number
  // Financier (gate PAYEE)
  quantite?: number
  prix_unitaire_cts?: number
  taux_tva_centieme?: number
  // Non-financier (gate PAYEE non applicable)
  marque?: string
  categorie_code?: string
}

export interface ValidatedLignesBatchEditRequest {
  updates: ValidatedLigneEdit[]
}

export interface ValidatedInvoiceMetaEditRequest {
  numero_facture?: string
  date_facture?: string
  vendor_code?: string
}

export interface ValidatedLinesEditWarning {
  code: string
  idx: number
  produit_id?: number
  ean?: string | null
  stock_before?: number
  stock_after?: number
  delta_requested?: number
  delta_applied?: number
  message: string
}

export interface ValidatedLignesEditResponse extends EtlImportDetail {
  warnings: ValidatedLinesEditWarning[]
}

export interface LigneAddRequest {
  designation: string
  unite_base?: string
  source_fournisseur?: string
  ean?: string
  marque?: string
  conditionnement?: string
  categorie_code?: string
  quantite?: number
  prix_unitaire_cts?: number
  taux_tva_centieme?: number
}

// ── Categories ───────────────────────────────────────────────────────────────

export interface CategoryItem {
  code: string
  label: string
}

export interface CategoryGroup {
  group: string
  label: string
  items: CategoryItem[]
}

export interface CategoriesResponse {
  groups: CategoryGroup[]
  total: number
}

// ── Validation helpers ───────────────────────────────────────────────────────

export type LigneValidationStatus = 'valid' | 'warning' | 'error'

/** Niveau de problème par champ : null = OK */
export type FieldAlert = 'error' | 'warning' | null

export interface LigneFieldAlerts {
  ean: FieldAlert
  categorie_code: FieldAlert
  marque: FieldAlert
  quantite: FieldAlert
  prix_unitaire_cts: FieldAlert
}

/** Message court par champ pour guider l'opérateur */
export const FIELD_ACTION_LABELS: Record<keyof LigneFieldAlerts, string> = {
  ean: '+ EAN',
  categorie_code: '+ Cat.',
  marque: '+ Marque',
  quantite: '+ Qte',
  prix_unitaire_cts: '+ Prix',
}

export interface LigneValidation {
  status: LigneValidationStatus
  fields: LigneFieldAlerts
  issueCount: number
}

export function validateLigne(ligne: LigneFactureRead): LigneValidation {
  const fields: LigneFieldAlerts = {
    ean: !ligne.ean ? 'warning' : null,
    categorie_code: (!ligne.categorie_code || ligne.categorie_code === 'AUTRE') ? 'warning' : null,
    marque: !ligne.marque ? 'warning' : null,
    quantite: (!ligne.quantite || ligne.quantite <= 0) ? 'error' : null,
    prix_unitaire_cts: (!ligne.prix_unitaire_cts || ligne.prix_unitaire_cts <= 0) ? 'error' : null,
  }

  const hasError = Object.values(fields).some(v => v === 'error')
  const hasWarning = Object.values(fields).some(v => v === 'warning')
  const issueCount = Object.values(fields).filter(v => v !== null).length

  const status: LigneValidationStatus =
    hasError ? 'error' : hasWarning ? 'warning' : 'valid'

  return { status, fields, issueCount }
}

File: app/schemas/catalogue/etl_import.py
"""Schemas Pydantic — ETL Imports (queue factures fournisseur ADR-25).

Couvre :
  - EtlImportRead : vue liste (queue preview + historique)
  - EtlImportDetail : vue détaillée avec lignes
  - EtlImportUpdateRequest : modification preview par opérateur
  - LigneFactureUpdate : modification d'une ligne par l'opérateur
  - FactureUploadResponse : réponse après upload
  - CategoryGroup : catégories groupées pour dropdown
"""
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class EtlImportRead(BaseModel):
    """Vue liste d'un import ETL (queue et historique)."""

    id: int
    fournisseur_id: Optional[int] = None
    fichier_source: Optional[str] = None
    fichier_path: Optional[str] = None
    statut: str
    nb_lignes_total: Optional[int] = None
    nb_lignes_ok: int = 0
    nb_lignes_conflit: int = 0
    nb_lignes_erreur: int = 0
    erreur_detail: Optional[str] = None

    # Métadonnées facture (ADR-25)
    numero_facture: Optional[str] = None
    date_facture: Optional[date] = None
    montant_ht_total: Optional[int] = None
    montant_tva_total: Optional[int] = None
    montant_ttc_total: Optional[int] = None
    vendor_code: Optional[str] = None
    quality_score: Optional[int] = None
    ecart_reconciliation: Optional[Decimal] = None

    # Routing multi-tenant (TAIYAT INCONTOURNABLE/NOUTAM)
    client_name: Optional[str] = None
    target_tenant_id: Optional[int] = None

    # Progression + Revert
    validation_step: Optional[str] = None
    reverted_at: Optional[datetime] = None
    reverted_by_id: Optional[int] = None

    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EtlImportListResponse(BaseModel):
    """Réponse paginée de la liste des imports."""

    items: list[EtlImportRead]
    total: int


class LigneFactureRead(BaseModel):
    """Ligne produit parsée (preview facture)."""

    idx: int = 0
    designation: str
    unite_base: str
    source_fournisseur: str
    ean: Optional[str] = None
    marque: Optional[str] = None
    conditionnement: Optional[str] = None
    categorie_code: Optional[str] = None
    quantite: Optional[float] = None
    prix_unitaire_cts: Optional[int] = None
    montant_ht_cts: Optional[int] = None
    montant_ttc_cts: Optional[int] = None
    taux_tva_centieme: Optional[int] = None
    designation_raw: Optional[str] = None
    article_fournisseur: Optional[str] = None
    degre_alcool: Optional[float] = None
    contenant: Optional[str] = None
    volume_unitaire_ml: Optional[int] = None
    colisage: Optional[int] = None

    # Coordonnées PDF et confiance (Phase 2)
    page_number: Optional[int] = None
    y_position: Optional[float] = None
    confidence_score: Optional[int] = None

    # Produits similaires et prix catalogue (Phase 2+3)
    similar_products: Optional[list] = None
    prix_catalogue_actuel_cts: Optional[int] = None

    # Auto-fill traçable (S1-S5) — métadata par champ auto-appliqué
    auto_applied_fields: Optional[dict] = None


class LigneFactureUpdate(BaseModel):
    """Modification d'une ou plusieurs lignes en PREVIEW."""

    idx: int = Field(..., ge=0, description="Index de la ligne dans le tableau")
    designation: Optional[str] = None
    ean: Optional[str] = None
    marque: Optional[str] = None
    conditionnement: Optional[str] = None
    categorie_code: Optional[str] = None
    quantite: Optional[float] = Field(None, ge=0)
    prix_unitaire_cts: Optional[int] = Field(None, ge=0)
    taux_tva_centieme: Optional[int] = Field(None, ge=0)
    # S5 — révocation d'auto-fill : le frontend peut renvoyer le dict tel
    # qu'il le veut persisté (ex: sans la clé du champ révoqué).
    auto_applied_fields: Optional[dict] = None


class LignesBatchUpdateRequest(BaseModel):
    """Batch update de lignes parsées."""

    updates: list[LigneFactureUpdate] = Field(..., min_length=1)


class LigneAddRequest(BaseModel):
    """Ajout manuel d'une ligne produit."""

    designation: str = Field(..., min_length=1, max_length=200)
    unite_base: str = Field(default="PIECE")
    source_fournisseur: str = Field(default="MANUAL")
    ean: Optional[str] = None
    marque: Optional[str] = None
    conditionnement: Optional[str] = None
    categorie_code: Optional[str] = None
    quantite: Optional[float] = Field(None, ge=0)
    prix_unitaire_cts: Optional[int] = Field(None, ge=0)
    taux_tva_centieme: Optional[int] = Field(None, ge=0)


class EtlImportDetail(EtlImportRead):
    """Vue détaillée d'un import avec les lignes parsées."""

    lignes: list[LigneFactureRead] = Field(default_factory=list)
    montant_ht_calcule: Optional[int] = None
    montant_ttc_calcule: Optional[int] = None


class ValidatedLignesEditResponse(EtlImportDetail):
    """Réponse PATCH validated-lignes — EtlImportDetail + avertissements downstream.

    `warnings` contient les alertes non bloquantes (ex: STOCK_NEGATIVE quand
    le stock est passé sous zéro suite au delta). Le client peut les afficher
    en toast pour informer l'opérateur.
    """

    warnings: list[dict] = Field(default_factory=list)


class ValidatedLigneEdit(BaseModel):
    """Édition d'une ligne d'un import VALIDATED.

    Champs financiers (P1) : quantite, prix_unitaire_cts, taux_tva_centieme.
    Champs non-financiers (P2) : marque, categorie_code.
    EAN et designation sont interdits post-validation (reparenting produit ingérable).

    Gate facture PAYEE : s'applique uniquement si au moins un champ financier
    est présent dans l'update (les modifs non-financières restent autorisées
    même sur une facture payée).
    """

    idx: int = Field(..., ge=0, description="Index ligne dans lignes_data")
    # Financier
    quantite: Optional[float] = Field(None, ge=0)
    prix_unitaire_cts: Optional[int] = Field(None, ge=0)
    taux_tva_centieme: Optional[int] = Field(None, ge=0)
    # Non-financier (P2)
    marque: Optional[str] = Field(None, max_length=100)
    categorie_code: Optional[str] = Field(None, max_length=50)


class ValidatedLignesBatchEditRequest(BaseModel):
    """Batch édition de lignes sur un import VALIDATED."""

    updates: list[ValidatedLigneEdit] = Field(..., min_length=1)


class ValidatedInvoiceMetaEditRequest(BaseModel):
    """Édition des métadonnées facture d'un import VALIDATED (P2).

    Répercute les champs sur l'EtlImport ET sur FinanceInvoice liée
    (reference / date_facture). Non-financier → pas de gate PAYEE.
    """

    numero_facture: Optional[str] = Field(None, max_length=100)
    date_facture: Optional[date] = None
    vendor_code: Optional[str] = Field(None, max_length=50)


class EtlImportUpdateRequest(BaseModel):
    """Modification d'un import en PREVIEW par l'opérateur."""

    numero_facture: Optional[str] = None
    date_facture: Optional[date] = None
    montant_ht_total: Optional[int] = Field(None, ge=0)
    montant_tva_total: Optional[int] = Field(None, ge=0)
    montant_ttc_total: Optional[int] = Field(None, ge=0)


class FactureUploadResponse(BaseModel):
    """Réponse après upload d'un fichier fournisseur."""

    etl_import_id: int
    statut: str
    nb_lignes: int
    numero_facture: Optional[str] = None
    vendor_code: Optional[str] = None
    quality_score: Optional[int] = None


class CategoryItem(BaseModel):
    """Une catégorie alimentaire."""

    code: str
    label: str


class CategoryGroup(BaseModel):
    """Groupe de catégories pour dropdown."""

    group: str
    label: str
    items: list[CategoryItem]


class CategoriesResponse(BaseModel):
    """Liste groupée des catégories alimentaires."""

    groups: list[CategoryGroup]
    total: int


class RevertImportResponse(BaseModel):
    """Résultat d'un revert d'import ETL."""

    mouvements_annules: int
    invoice_annulee: bool
    prix_restaures: int


class DashboardQualityTrend(BaseModel):
    """Point de tendance qualité."""

    id: int
    import_date: Optional[date] = Field(None, alias="date")
    quality_score: Optional[int] = None
    vendor_code: Optional[str] = None
    nb_lignes_total: Optional[int] = None

    model_config = ConfigDict(populate_by_name=True)


class DashboardVendorStats(BaseModel):
    """Stats par fournisseur."""

    vendor_code: str
    total_imports: int
    avg_quality: Optional[float] = None
    recognition_rate: Optional[float] = None


class DashboardResponse(BaseModel):
    """Dashboard analytique ETL."""

    total_imports: int
    by_status: dict[str, int]
    avg_quality_global: Optional[float] = None
    quality_trend: list[DashboardQualityTrend]
    vendor_stats: list[DashboardVendorStats]
    pending_conflicts: int
    avg_correction_time_sec: Optional[float] = None
